### PR TITLE
10141: Avoid trying to call cancel() on the timeout if it is None.

### PR DESCRIPTION
**Pull request by wiml**
Also, try to clear out callback variables a little earlier to avoid the possibility of them being callback()'d twice.

This fixes 10141, and might also fix [#9494](https://twistedmatrix.com/trac/ticket/9494); I can't be sure without being able to replicate it, but the backtrace in the trac ticket is suggestive.

## Contributor Checklist:

* [X] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10141
* [X] I ran `tox -e lint`
* [X] I have created a newsfragment
* [X] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] I have added `twisted/twisted-contributors` teams to the PR `Reviewers`.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #1558 from twisted/wiml-namesTimeout-10141

Author: wiml
Reviewer: <github_username>, <github_usernames_if_more_reviewers>
Fixes: ticket:10141
Refs: ticket:9494 ticket:3428

Avoid trying to call cancel() on the timeout if it is None, and clean up
better during abnormal situations.
```

### CODE REVIEW

**Comment by wiml** [comment]
I think this is ready. The check failures are:

* Codecov.io says that the PR decreases overall test coverage, but I don't see how it could. If I look at the coverage report, about 2/3 of the time I get an authorization error from codecov and can't see anything; the other 1/3 of the time it reports coverage decreases in completely unrelated code, like the TLS tests. I think this is spurious.
* Codecov.io complains the diff test coverage isn't 100%. As far as I can tell, 100% of the diff *lines* are covered by tests; however, one defensive `if` statement is not fully exercised. I can't make both codecov and mypy happy here. I think the current code is more correct than if I didn't have the `if`.
* I frequently get one or two spurious failures in other parts of Twisted, *e.g.* in disttrial. They move around if I re-run the tests, and don't seem to be related to the changes in this PR. Sometimes I don't get any of these spurious failures. I assume they are an intermittent bug elsewhere in Twisted or an instability in the CI infra but I can't be sure. In any event I can't think of a way they could be connected to these AXFR fixes.


You'll also notice one TODO-skipped test; this is because these new unit tests also elicit the failure described in [trac #3428](https://twistedmatrix.com/trac/ticket/3428). I have another patch queued up to fix that, after this one is merged, but I want to get this fix merged so I can get on with my actual work.

**Comment by wsanchez** [comment]
You do need to test the code past where `connector` is `None`, or remove the defensive check if it's not necessary.

What's `mypy` complaining about there?

**Comment by wiml** [comment]
Ah, you're right. I was thinking of a similar class where `connector` is an ivar and therefore has to be typed `Optional`. But in this class it's threaded through the deferreds so it clearly won't be `None`.

**Comment by adiroiban** [comment]
@wiml thanks for the pr.

you  can ping me if you found random/spurious/flaky tests. My goal in life is to get a solid green for all tests from trunk :)

I see some tests failing , but not all failures are due to parallel jobs.

For example this fails even without parallel
```
$ trial -u twisted.trial._dist.test.test_disttrial
```

but pass with

```
$ trial twisted.trial._dist.test.test_disttrial
```

-----------

coverage reporting should be broken, unless this is solved/merged https://github.com/twisted/twisted/pull/1574

I have no idea how you got a green build for https://github.com/twisted/twisted/pull/1558/commits/79181664e0eccaab91ec5f31898ea3e0eaa9fc6b

**Comment by wiml** [comment]
@adiroiban In my most recent push c68b1db there's a spurious-looking failure in one of the pypy runs, which looks like it might simply have timed out after half an hour.

Regarding that one green commit I'm not sure either, since codecov only occasionally lets me see the results of runs. :man_shrugging: One thing I have noticed, though, is that `codecov/project` usually succeeds/fails long before the various test runs do, so perhaps its pass/fail depends on which test runs have completed when it gathers coverage data?

**Comment by adiroiban** [comment]
I am working on PYPY failures at https://github.com/twisted/twisted/pull/1543

For codecov/project yes, you get a report as soon as the first report is sent  by a job... and then it is updated.
I can look at configuring that to delay the report.

**Comment by wiml** [comment]
In that case I think I'll wait for the CI issues to get sorted out on trunk and then come back to this fix.

File: src/twisted/names/test/test_names.py [hunk 1]
```diff
@@ -472,6 +506,65 @@ def test_zoneTransferConnectionFails(self):
...
+            EasilyStartledProtocol,
+            DomainError,
+        )
+
+    test_zoneTransferConnectionDrops.todo = "Bug #3428"  # type: ignore[attr-defined]
+
+    def test_zoneTransferTimesOut(self) -> defer.Deferred:
+        """
+        An AXFR TCP transfer which never receives a response from the
+        server should cleanly time out.
```
Comment (adiroiban) [inline src/twisted/names/test/test_names.py]: What does  "cleanly time out" mean? 

What is the target / subject / system under test here?

Would this be a better description of this test.

```
L{Resolver,lookupZone} returns a deferred which fails with L{DNSQueryTimeoutError}
when the remote DNS server is not responding before the timeout interval.
```

**Comment by adiroiban** [comment]
I don't think that you should block on the random CI failures. Just make sure the code from this PR is rock solid and I am sure that it will be approved.

There will be always CI failures, as keeping a CI happy is a full time job description :)

You can try to review #1543 :)

I have triggered an update on this branch to get latest CI fixes.

I am dedicating all my current energy to improving the overall Twisted CI dev experience and I don't have the energy required to do a review :(

File: src/twisted/names/test/test_names.py (hunk 1 above)
Comment (wiml) [reply src/twisted/names/test/test_names.py]: That sounds good, I've clarified the docstring.

**Comment by wiml** [comment]
Finally a green CI check (and faster, too) — thanks for all your work, Adi!

**Comment by wiml** [comment]
Ping?

File: src/twisted/names/test/test_names.py [hunk 2]
```diff
@@ -469,6 +502,66 @@ def test_zoneTransferConnectionFails(self):
...
         )
 
+    def _testHelper_zoneTransferFailure(
+        self,
+        protocolClass: Type[AccumulatingProtocol],
+        expectedFailure: Type[Exception],
+        **kwargs: Any,
+    ) -> defer.Deferred:
+        factory = RecordingServerFactory.forProtocol(protocolClass)
+        nonresponsiveTCP = reactor.listenTCP(0, factory, interface="127.0.0.1")  # type: ignore[attr-defined]
```
Comment (twm) [inline src/twisted/names/test/test_names.py:512]: Could this be done with [`MemoryReactor`](https://twistedmatrix.com/documents/current/api/twisted.internet.testing.MemoryReactor.html) instead of the real reactor? There is a version that provides `IReactorUDP` in `twisted.names.test.test_util`.

File: src/twisted/names/test/test_names.py [hunk 3]
```diff
@@ -469,6 +502,66 @@ def test_zoneTransferConnectionFails(self):
...
+            resolver.lookupZone("test-domain.com", **kwargs), expectedFailure
+        )
+
+        def _waitAnd(r, fn, *args):
+            # We need to wait a moment before calling _furtherChecks
+            # for the TCP closure to work its way through the stack
+            # and show up as accumulator.closed==True.
+            d = defer.Deferred()
+            d.addCallback(fn, *args)
+            reactor.callLater(0.01, d.callback, r)
```
Comment (twm) [inline src/twisted/names/test/test_names.py:526]: Blind sleeps lead to flaky tests. :grimacing: This wouldn't be necessary with a memory reactor.

File: src/twisted/names/test/test_names.py [hunk 4]
```diff
@@ -469,6 +502,66 @@ def test_zoneTransferConnectionFails(self):
...
+        """
+        If the TCP connection initiated by L{Resolver.lookupZone} is lost before any
+        response is received, the deferred will errback.
+        """
+        return self._testHelper_zoneTransferFailure(
+            EasilyStartledProtocol,
+            DomainError,
+        )
+
+    test_zoneTransferConnectionDrops.todo = "Bug #3428"  # type: ignore[attr-defined]
```
Comment (twm) [inline src/twisted/names/test/test_names.py:553]: _Technically_ we're not supposed to have TODOs in trunk, but I'd allow it.

**Review by twm — CHANGES_REQUESTED** [review]
The fix here looks good to me! However I'm concerned about the use of the real reactor in the test, particularly the sleep. Listening on ports like this tends to cause flakiness on the scale of Twisted's suite, so we should avoid it whenever possible, which looks quite possible here.

File: src/twisted/names/newsfragments/10141.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+t.n.client.Resolver will no longer throw an AttributeError if a zone transfer times out.
```
Comment (exarkun) [inline src/twisted/names/newsfragments/10141.bugfix:1]: Please expand the abbreviations and use "raise" instead of "throw".

File: src/twisted/names/test/test_names.py [hunk 6]
```diff
@@ -469,6 +502,66 @@ def test_zoneTransferConnectionFails(self):
             resolver.lookupZone("impossible.invalid"), error.DNSLookupError
         )
 
+    def _testHelper_zoneTransferFailure(
```
Comment (exarkun) [inline src/twisted/names/test/test_names.py:505]: This should be `_testHelperZoneTransferFailure` because it is not dynamically dispatched.

File: src/twisted/names/test/test_names.py [hunk 7]
```diff
@@ -469,6 +502,66 @@ def test_zoneTransferConnectionFails(self):
...
+        self,
+        protocolClass: Type[AccumulatingProtocol],
+        expectedFailure: Type[Exception],
+        **kwargs: Any,
```
Comment (exarkun) [inline src/twisted/names/test/test_names.py:509]: I think that `object` should be preferred over `Any`?

File: src/twisted/names/test/test_names.py (hunk 2 above)
Comment (exarkun) [reply src/twisted/names/test/test_names.py:512]: `listenTCP(0, ...)` will fail sometimes with the real reactor which is further reason to use a test double instead.

File: src/twisted/names/test/test_names.py (hunk 4 above)
Comment (exarkun) [reply src/twisted/names/test/test_names.py:553]: Why? :slightly_smiling_face: It could go on the ticket or in a new branch associated with the ticket...
